lib/RouterMixin.js

```diff
@@ -83,11 +83,36 @@ RouterMixin = {
         }
 
         self._historyListener = function (e) {
+            debug('history listener invoked', e, url, self.state.route.url);
+
             if (context) {
+                var state = self.state || {};
                 var url = self._history.getUrl();
-                debug('history listener invoked', e, url, self.state.route.url);
-                if (url !== self.state.route.url) {
-                    context.executeAction(navigateAction, {type: TYPE_POPSTATE, url: url, params: (e.state && e.state.params)});
+                var currentUrl = state.route.url;
+                var route = state.route || {};
+                var onBeforeUnloadText = typeof window.onbeforeunload === 'function' ? window.onbeforeunload() : '';
+                var confirmResult = onBeforeUnloadText ? window.confirm(onBeforeUnloadText) : true;
+                var nav = route.navigate || {};
+                var navParams = nav.params || {};
+                var enableScroll = self._enableScroll && nav.preserveScrollPosition;
+                var historyState = {
+                    params: (nav.params || {}),
+                    scroll: {
+                        x: (enableScroll ? window.scrollX : 0),
+                        y: (enableScroll ? window.scrollY : 0)
+                    }
+                };
+                var pageTitle = navParams.pageTitle || null;
+
+                if (!confirmResult && currentUrl) {
+                    // Pushes the previous history state back on top to set the correct url
+                    self._history.pushState(historyState, pageTitle, currentUrl);
+                } else {
+                    window.onbeforeunload = null;
```

Comment:
The `onbeforeunload` handler is supposed to unique per page (since not all pages will need it).  This just makes sure to remove it so that other pages aren't affected.

The app developer will implement the `onbeforeunload` handler for each page that they need it.